internal/logger/command.go

```diff
@@ -0,0 +1,181 @@
+package logger
+
+import (
+	"strconv"
+	"time"
+
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/bson/bsontype"
+	"go.mongodb.org/mongo-driver/bson/primitive"
+)
+
+// DefaultMaxDocumentLength is the default maximum number of bytes that can be
+// logged for a stringified BSON document.
+const DefaultMaxDocumentLength = 1000
+
+// TruncationSuffix are trailling ellipsis "..." appended to a message to
+// indicate to the user that truncation occurred. This constant does not count
+// toward the max document length.
+const TruncationSuffix = "..."
+
+const (
+	CommandMessageFailedDefault    = "Command failed"
+	CommandMessageStartedDefault   = "Command started"
+	CommandMessageSucceededDefault = "Command succeeded"
+
+	// CommandMessageDroppedDefault indicates that a the message was dropped
+	// likely due to a full buffer. It is not an indication that the command
+	// failed.
+	CommandMessageDroppedDefault = "Command message dropped"
+)
+
+type CommandMessage struct {
+	DriverConnectionID int32
+	MessageLiteral     string
+	Name               string
+	OperationID        int32
+	RequestID          int64
+	ServerConnectionID *int32
+	ServerHost         string
+	ServerPort         string
+	ServiceID          *primitive.ObjectID
+}
+
+func (*CommandMessage) Component() Component {
+	return ComponentCommand
+}
+
+func (msg *CommandMessage) Message() string {
+	return msg.MessageLiteral
+}
+
+func serializeKeysAndValues(msg CommandMessage) ([]interface{}, error) {
```

Comment:
What's the reason to return `[]interface{}` instead of `map[string]interface{}` or even `bson.M`? It seems like later on we'd have to convert the list of ("string", value) pairs to `map[string]interface{}` anyway in `osSink` at least:

```go
for i := 0; i < len(keysAndValues); i += 2 {
 	kvMap[keysAndValues[i].(string)] = keysAndValues[i+1]
}
```
Using `map[string]interface{}` would avoid this eventual conversion and allow easier setting of variables below through map syntax.

Reply:
The map thing you mention is the default sink using the standard library. In general, the LogSink methods will pass the serialized arguments as a slice of empty interfaces of the form "key1, value1, key2, value2, ...". This is more relaxed and standard across most logging libraries, and is the way logr defines their LogSink interface.